crates/holochain/src/core/workflow/app_validation_workflow/validation_dependencies.rs

```diff
@@ -0,0 +1,288 @@
+use std::{
+    collections::{hash_map::Entry, HashMap, HashSet},
+    time::{Duration, Instant},
+};
+
+use holo_hash::{AnyDhtHash, DhtOpHash};
+use holochain_types::dht_op::DhtOpHashed;
+
+/// In-memory struct to keep track of missing DHT hashes, which DhtOp depends on them
+/// and when they were started being fetched.
+pub struct ValidationDependencies {
+    /// Missing hashes that are being fetched, along with a set of DhtOps that depend
+    /// on the hash and the last Instant a fetch was attempted.
+    missing_hashes: HashMap<AnyDhtHash, (HashSet<DhtOpHash>, Instant)>,
+}
+
+impl Default for ValidationDependencies {
+    fn default() -> Self {
+        ValidationDependencies::new()
+    }
+}
+
+impl ValidationDependencies {
+    const FETCH_TIMEOUT: Duration = Duration::from_secs(60);
+
+    pub fn new() -> Self {
+        Self {
+            missing_hashes: HashMap::new(),
+        }
+    }
+
+    // returns true if this is a new missing hash
+    pub fn insert_missing_hash_for_op(&mut self, hash: AnyDhtHash, dht_op_hash: DhtOpHash) -> bool {
+        if let Entry::Vacant(entry) = self.missing_hashes.entry(hash.clone()) {
+            let mut dht_op_hashes = HashSet::new();
+            dht_op_hashes.insert(dht_op_hash);
+            entry.insert((dht_op_hashes, Instant::now()));
+            true
+        } else {
+            self.missing_hashes
+                .entry(hash)
+                .and_modify(|(dht_op_hashes, _)| {
+                    dht_op_hashes.insert(dht_op_hash.clone());
+                });
+            false
+        }
+    }
+
+    pub fn remove_missing_hash(&mut self, hash: &AnyDhtHash) {
+        self.missing_hashes.remove(hash);
+    }
+
+    pub fn get_missing_hashes(&self) -> &HashMap<AnyDhtHash, (HashSet<DhtOpHash>, Instant)> {
+        &self.missing_hashes
```

Comment:
Gives up internal state, what is this used for that shouldn't be a function on this type?

Reply:
That was a perhaps lazy fix to tests accessing the internal state which was possible while this struct was still inside of the app validation workflow module. Is it preferable to change the field's scope to `pub(super)` or something like that?

Reply:
Oh it's fair enough for testing but then maybe put it behind a `#[cfg(test)]`. I was wondering if that's the only place it was used from

Reply:
Done